lib/plug/adapters/cowboy2.ex

```diff
@@ -0,0 +1,336 @@
+
+  ## Examples
+
+  Assuming your Plug module is named `MyApp` you can add it to your
+  supervision tree by using this function:
+
+      children = [
+        {Plug.Adapters.Cowboy2, scheme: :http, plug: MyApp, options: [port: 4040]}
+      ]
+
+      Supervisor.start_link(children, strategy: :one_for_one)
+
+  """
+  def child_spec(opts) do
+    scheme = Keyword.fetch!(opts, :scheme)
+    cowboy_opts = Keyword.fetch!(opts, :options)
+
+    {plug, plug_opts} =
+      case Keyword.fetch!(opts, :plug) do
+        {_, _} = tuple -> tuple
+        plug -> {plug, []}
+      end
+
+    {id, start, restart, shutdown, type, modules} =
+      child_spec(scheme, plug, plug_opts, cowboy_opts)
+
+    %{id: id, start: start, restart: restart, shutdown: shutdown, type: type, modules: modules}
+  end
+
+  defp child_spec(scheme, plug, opts, cowboy_options) do
+    [ref, trans_opts, proto_opts] = args(scheme, plug, opts, cowboy_options)
+
+    cowboy_function =
+      case scheme do
+        :http -> :start_clear
+        :https -> :start_tls
+      end
+
+    cowboy_args = [ref, trans_opts, proto_opts]
+
+    {
+      {:ranch_listener_sup, ref},
+      {:cowboy, cowboy_function, cowboy_args},
+      :permanent,
+      :infinity,
+      :supervisor,
+      [:ranch_listener_sup]
+    }
+  end
+
+
+  ## Helpers
+
+  @protocol_options [:timeout, :compress, :stream_handlers]
+
+  defp run(scheme, plug, opts, cowboy_options) do
+    case Application.ensure_all_started(:cowboy) do
+      {:ok, _} ->
+        :ok
+
+      {:error, {:cowboy, _}} ->
+        raise "could not start the Cowboy application. Please ensure it is listed as a dependency in your mix.exs"
+    end
+
+    start =
+      case scheme do
+        :http -> :start_clear
+        :https -> :start_tls
+        other -> :erlang.error({:badarg, [other]})
+      end
+
+    apply(:cowboy, start, args(scheme, plug, opts, cowboy_options))
+  end
+
+  defp set_compress(cowboy_options) do
+    compress = Keyword.get(cowboy_options, :compress)
+    stream_handlers = Keyword.get(cowboy_options, :stream_handlers)
+
+    case {compress, stream_handlers} do
+      {true, nil} ->
+        Keyword.put_new(cowboy_options, :stream_handlers, [
+          :cowboy_compress_h,
+          Plug.Adapters.Cowboy2.BadResponseCheck,
+          :cowboy_stream_h
+        ])
+
+      {true, _} ->
+        raise "Cannot set both compress and stream_handlers at once. " <>
+                "If you wish to set compress, please add `:cowboy_compress_h` to your stream handlers."
+
+      _ ->
+        cowboy_options
+    end
+  end
+
+  defp normalize_cowboy_options(cowboy_options, :http) do
+    Keyword.put_new(cowboy_options, :port, 4000)
+  end
+
+  defp normalize_cowboy_options(cowboy_options, :https) do
+    assert_ssl_options(cowboy_options)
+    cowboy_options = Keyword.put_new(cowboy_options, :port, 4040)
+
+    cowboy_options =
+      Enum.reduce(
+        [:keyfile, :certfile, :cacertfile, :dhfile],
```

Comment:
It would be nice to also move this to a module attribute or to its own variable so everything fits in a single line.